base/math.jl

```diff
@@ -536,38 +536,57 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+    #= Keep specialized two argument version, because it is faster
+    than generic varargs version below =#
+    ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
     if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
+        if isinf(x) || isinf(y)
+            return oftype(rr, Inf)
+        else
+            return oftype(rr, r)
+        end
     else
         return rr
-    end
+   	end
 end
-
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+    N == 0 && error("invalid hypot() call without arguments")
+    # speculatively try fast naive approach
+    s = sum(abs2.(float.(x)))
+    r = sqrt(s)
+    if isnan(s)
+        return any(isinf, x) ? oftype(r, Inf) : s # IEEE 754
+    elseif isinf(s) || s ≤ _floatmin(s) * (N-1) # FIXME: revisit when #30387 is resolved
+        # if overflow/underflow, try normalization
+        ma = maximum(abs, x)
```

Comment:
I think you need `float(maximum(abs, x))` here.

* Then you can eliminate the `oftype` calls, so that you no longer need `r` to be computed (i.e. you eliminate the `sqrt` call).

* It looks like your current code can overflow for rationals, since `x ./ ma` will produce another rational that can overflow in the `abs2`.  In particular, consider `let x = typemax(Int); @test hypot(x//1, x//1, x//1) ≈ x*√3; end`

Reply:
Thanks! I added this test case